base_layer/core/src/chain_storage/blockchain_database.rs

```
@@ -1926,11 +2005,21 @@ fn reorganize_chain<T: BlockchainBackend>(
             backend.write(txn)?;
 
             info!(target: LOG_TARGET, "Restoring previous chain after failed reorg.");
-            restore_reorged_chain(backend, fork_hash, removed_blocks, consensus)?;
+            restore_reorged_chain(backend, fork_hash, removed_blocks, consensus, smt.clone())?;
             return Err(e.into());
         }
 
-        insert_best_block(&mut txn, block.clone(), consensus)?;
+        if let Err(e) = insert_best_block(&mut txn, block.clone(), consensus, smt.clone()) {
+            let mut write_smt = smt.write().map_err(|e| {
+                error!(
+                    target: LOG_TARGET,
+                    "An attempt to get a write lock on the smt failed. {:?}", e
+                );
+                ChainStorageError::AccessError("write lock on smt".into())
+            })?;
+            *write_smt = backend.calculate_tip_smt()?;
+            return Err(e);
+        }
```

I count 3x places in the code where this pattern is employed, where the SMT needs to be recalculated if some underlying error happens, If this function (`calculate_tip_smt`) errors we cannot recover. 

Maybe at this point the SMT should be wiped, and we should have a wrapper function that checks to see if the SMT is empty and try to calculate it whenever we want to use it.

---

if calculate_tip_smt fails, you entire db is stuffed, and you need to sync from scratch. Aka wipe the node. So I don't see the need to add a helper function to wipe it and only let it fail again. This replaces the SMT in memory.

---

It is better to have no SMT than to have one that is not in sync with the db. The calculate function could fail due to other reasons, not only due to bad lmdb data, for example the OS could be busy swapping memory to disk for a long time and impair other read operations.

---

for that a restart of the node is better anyway

---

Not resolved